sql/core/src/test/scala/org/apache/spark/sql/DatasetAggregatorSuite.scala

```diff
@@ -206,4 +216,12 @@ class DatasetAggregatorSuite extends QueryTest with SharedSQLContext {
       ds.groupByKey(_.b).agg(ComplexBufferAgg.toColumn),
       ("one", 1), ("two", 1))
   }
+
+  test("SPARK-12555 - data is not corrupted after input data is reordered") {
```

Comment:
@cloud-fan  Exactly, and this generates the erroneous result as described in the jira if running on 1.6, but works as expected in 2.0. I want to make sure we enhance the test coverage here, to avoid regressions in the future, so we can close the jira issue after we get this test scenario into the 2.0 codebase.